### ISSUE TITLE
Expand GBM to include other predictors

### ISSUE THREAD

**Issue by tomschenkjr** [Chicago/e-coli-beach-predictions#62]
Being led by @beckeroobonsai -- expand GBM by including additional predictors (e.g., weather) for the applicable periods.

**Comment by tomschenkjr** [comment]
Initial model showed tremendous progress and only using previous geometric mean from 2002-2016:

![image](https://cloud.githubusercontent.com/assets/1639121/13273104/a14e7e96-da66-11e5-8a1c-d98a0a664840.png)

![image](https://cloud.githubusercontent.com/assets/1639121/13273117/bc4d628e-da66-11e5-87ab-de3d265f20db.png)

GradientBoostingClassifier vs LogisticRegression. Model variables are :[ 'Beach', 'Month', 'Weekday', '7_day_moving_avg', 'PriorReading',  'Montrose_prior', 'Calumet_prior', '63rd_prior', 'Rainbow_prior', 'Ohio_prior', '31st_prior', '57th_prior', 'South Shore_prior', '12th_prior', 'Ostermann_prior', 'Loyola_prior', 'Oak_prior',
      'Foster_prior', 'North Ave_prior', 'Albion_prior', 'Jarvis_prior',
      'Rogers_prior', 'Howard_prior', 'Juneway_prior', '39th_prior',
      'Pratt_prior', 'city_mean']

Next iteration of models should include additional variables.

**Comment by kbrose** [comment]
Wanted to copy the rest of the description of the model from slack:

> I built a model that only relies on Ecoli geometric mean readings.  That let me use many more years of data than possible if I merged on to weather or Drek readings.   Instead, I merge on the data from other beaches from one day prior to the current.  If that is missing due to it being Monday or some other reason, then the avg of previous 7 days (not 7 readings) is used.  If there is a NaN resulting from some beach not being sampled on a day where others are sampled, the NaN was replaced with the citywide mean for that day.    The Gradient Boosting Model allows for deep interactions between variables to be taken into account.  It performs much better than a simple Logistic Regression model.  The hyperparameters I used were:  n_estimators=1000, learning_rate=0.05, max_depth=6.

**Comment by kbrose** [comment]
After using data from 2014 and earlier as training and data from 2015 as testing, the model performance drops. Need to investigate what might have caused this. @ChrisProkop suggests that perhaps we should reserve 2015 data as a final testing set, and only work with pre-2015 data for a while to avoid potential over-fitting (see also #15).

![roc_curve](https://cloud.githubusercontent.com/assets/4053170/13486482/2f2de898-e0d7-11e5-84ae-0dc9dc1d9116.png)

![precision_recall](https://cloud.githubusercontent.com/assets/4053170/13486485/33edb25a-e0d7-11e5-9602-0cda7cbecbc4.png)

**Comment by kbrose** [comment]
@beckeroobonsai suggested investigating performance of this GBM model on a beach-by-beach basis to see if performance declined only in some beaches, generally decreased in all beaches, etc.

What is the best course of action with respect to fear of over-fitting by not having a reserved test set?

**Comment by ChrisProkop** [comment]
Some thoughts:

1) We need to be more judicious with how we handle the final validation set.  2015 shouldn't be looked at all yet, until we've done our due diligence in building a good model.  Just looking it and using it to inform your model building produces somewhat of the same effect as over-fitting the model through other means; you're essential building the model to predict well on 2015 data.

2) ROC curves and PR curves aren't the only good plots to use.  Univariate plots can help inform both what variables are useful for the model, and when made on data that wasn't used to to train can show whether the issue is an across-the-board shit or genuinely different behavior.

3) We should probably keep 2015 as an out-of-time validation set that we don't look at until we've spent a lot of time building models, then divide the rest of the day into a training set and a tuning set.  The tuning set is going to be more vital for a GLM, I think, but there's still some value into randomly dividing it into 80-20 or 70-30 sets.

**Comment by ChrisProkop** [comment]
I don't know anything at all about what is easy to do in Python GBMs and what isn't.  Is there a way to do the equivalent of a gbm.perf plot from R?

We can also do a Grid Search to see whether the GBM parameters are sufficient.

http://scikit-learn.org/stable/modules/grid_search.html

**Comment by ChrisProkop** [comment]
People mentioned some anti-bird methods being used over the years (chasing them away, or covering trash cans), which could have a considerable effect.  Is it possible to get some info on what the city has done for this?

**Comment by ChrisProkop** [comment]
Also, the reason why the ROC curves look so great when we have a random testing set is that, because we have prior days information in the data we are building off of, the training set technically has a (partial)? answer for what the values were during the random days taken out for testing.

Maybe.  I'm not entirely clear, but a random holdout definitely seems like there could be data leakage somehow.

**Comment by fgregg** [comment]
An alternative explanation is that 2015 was a different year than previous years. In other words, the test set is not drawn from the same distribution as the training set. If that's true, then you are in more trouble.

What happens to the ROC if we pull out a different years for the test set?

@ChrisProkop is sounds like you are saying that there's an auto-correlation between a beach's measure and what was going on the next day. But, the previous day's measure is explicitly in the model. The auto-correlation should be being exploited.

**Comment by ChrisProkop** [comment]
@fgregg Yes, we do need to utilize the previous day's information.  However, my concern is if the training  set and test set are odd & even days, respectively, as a worst-case-scenario.  Then your training set contains a full list of the answers to the testing set. I'm not convinced that is what is happening, but I need to think about it more.  GBMs can be kind of screwy, and there may be hard-to-conceptualize data leakage.   

Changes from year-to-year is likely a huge effect that we need to consider.   But thinking about it more, it might not be that the 2015 ROC plot is bad; it could be that that the other plots are just simply overfit.  Showing plots of performance on the same data that was used for training is kind of meaningless, because it is supposed to be awesome, and something like a GBM can specify the exact day-year-beach combinations (you wouldn't even need a tree-depth of 6 for that). 

That's kind of the value of using three different data partitations.

One to train on.
One on which to judge the quality of the model you've trained and to tune future runs of the model
One to open near the end of the project to validate if the model produced is general enough to be useful in the field.

**Comment by kbrose** [comment]
@ChrisProkop :

> I don't know anything at all about what is easy to do in Python GBMs and what isn't. Is there a way to do the equivalent of a gbm.perf plot from R?

I don't know if it exists out-of-box in the sklearn GBM, but it looks like the example in http://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_oob.html does what we want. Should be able to modify that code fairly easily.

---

@ChrisProkop @fgregg :

This weekend when I've got some time I want to get a GBM model on a branch for others to play around with, and also produce plots doing leave-one-year-out cross validation (ignoring 2015 completely, of course) to see if 2015's drop in performance is unique or not.

**Comment by kbrose** [comment]
Issue appears to not be unique to 2015:

![roc](https://cloud.githubusercontent.com/assets/4053170/13562558/9d4aca78-e3ff-11e5-89fb-a58c5cc379c0.png)
(Note that we only have performance for EPA model in 2015, and that our performance on 2015 is not shown here since it is reserved as a final testing set. Also keep in mind that this is using a similarly restrictive set of predictors as in the original GBM model above.)

Plot can be made by running `python gbm.py` inside `python_src` of the commit ca86dcc42.

(The PR curve was acting strange and it's too late for me to debug.)
